test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java

```diff
@@ -456,6 +461,218 @@ public void checkpointWriterInterleavedTest() throws Exception {
         }
     }
 
+    /** Test the CheckpointWriter write size limit.
+     *
+     * CheckpointWriter aggregates a batch of SMREntries into one
+     * CheckpointEntry. This test uses large-sized SMREntries to verify
+     * that batching will not violate the maxWriteSize limit.
+     */
+    @Test
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void checkpointWriterSizeLimitTest() throws Exception {
+        final String streamName = "mystream5";
+        final UUID streamId = CorfuRuntime.getStreamID(streamName);
+        final String keyPrefix = "a-prefix";
+        final int numKeys = 50;
+        final String author = "Me, myself, and I";
+        final String mutationSuffix = "_mutation_suffix";
+        // max write size set to 25MB
+        final int maxWriteSize = 25 << 20;
+
+        StreamingMap<String, String> m = instantiateStringMap(streamName);
+
+        for (int i = 0; i < numKeys; i++) {
+            // each entry is 1 MB
+            String payload = getRandomStringOfSize(1 << 20);
+            m.put(keyPrefix + i, payload);
+        }
+
+        // set the max write size to 25 MB
+        getRuntime().getParameters().setMaxWriteSize(maxWriteSize);
+        CheckpointWriter cpw = new CheckpointWriter(getRuntime(), streamId, author, (CorfuTable) m);
+        cpw.setSerializer(serializer);
+        // set a mutator to later verify that checkpointWriter has processed the entries
+        cpw.setValueMutator((l) -> (String) l + mutationSuffix);
+
+        // Write all CP data.
+        r.getObjectsView().TXBuild()
+                .type(TransactionType.SNAPSHOT)
+                .build()
+                .begin();
+        Token snapshot = TransactionalContext
+                .getCurrentContext()
+                .getSnapshotTimestamp();
+        try {
+            cpw.startCheckpoint(snapshot);
+            cpw.appendObjectState(m.entryStream());
+            cpw.finishCheckpoint();
+
+            // Instantiate new runtime & map.
+            setRuntime();
+            Map<String, String> m2 = instantiateStringMap(streamName);
+            for (int i = 0; i < numKeys; i++) {
+                assertThat(m2.get(keyPrefix + i)).describedAs("get " + i)
+                        .isEqualTo(m.get(keyPrefix + i) + mutationSuffix);
+            }
+        } finally {
+            r.getObjectsView().TXEnd();
+        }
+    }
+
+    /**
+     * CheckpointWriter has BATCH_THRESHOLD_PERCENTAGE which decreases the
+     * writeSizeLimit to (1-BATCH_THRESHOLD_PERCENTAGE) * writeSizeLimit.
+     * This test checks that a new checkpointEntry is created when the
+     * total size of SMREntries exceeds the decreased writeSizeLimit.
+     */
+    @Test
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void checkpointWriterSizeLimitMarginTest() throws Exception {
```

Comment:
Just to confirm, I'm guessing this test has compression disabled, right? otherwise there is no way to make sure we're triggering this scenario...

Reply:
The compression is enabled. This test is to verify that the new batching strategy works, and specifically verify that uncompressed size of a batch, which is going to be later compressed, is smaller than 98% * writeSizeLimit. Everything is uncompressed at this stage. We are not testing the size check that StreamView has.

Reply:
Oh when I discussed with Achin I thought the idea of this test was to test that when the size of the data + overhead exceeded the size limit we were fine... So we need to ensure that the data is exactly 98% such that when adding the overhead we don't get a WriteSizeLimitException, and the only way to control this is to have compression disabled, otherwise, 98% could turn in 60% + overhead is never near the boundary..

Reply:
Oh I finally see your point. So we generate some entries whose size add up to somewhere very close to 98% (still smaller) and verify that it doesn't trigger WriteSizeLimitException right?

Reply:
Correct. Well I would say exactly 98%, because we're assuming the 2% accounts for the overhead, so we want to test the border case, that is, the worst case when data + overhead could exceed the write size limit. Actually if you write this test and then change 0.98 to 1.0 and make your data 25 MB (or whichever your limit is) it should fail cause 100% + overhead > write size limit...

Reply:
Hi Anny, a new test case is added [here](https://github.com/CorfuDB/CorfuDB/pull/2875/files#diff-bd6de4f220c80c5a84d6c60f308074aac0a096efcefe915f124f15a932652cabR660). In this test, I disabled compression and changed batchThresholdPercentage to 100%, then generate entries whose total size is in range [maxWriteSize-100bytes, maxWriteSize) so that it could pass the first check and fail at the second check and triggers WriteSizeException due to the overhead of CheckpointEntry. Note that it's almost impossible to have total size of SMR entries exactly == 100% because we can't precisely decide the SMR entry size. Although we can provide a buf as `byte[] buf = new byte[size]`, the SMR entry that encapsulates this buf is unknown.